flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/BatchingStateChangeStore.java

```diff
@@ -0,0 +1,173 @@
+
+    @GuardedBy("scheduled")
+    private long scheduledSizeInBytes;
+
+    @GuardedBy("scheduled")
+    private Future<?> scheduledFuture = CompletableFuture.completedFuture(null);
+
+    private AtomicReference<Throwable> error = new AtomicReference<>(null);
+
+    BatchingStateChangeStore(
+            long persistDelayMs, long sizeThresholdBytes, StateChangeStore delegate) {
+        this(
+                persistDelayMs,
+                sizeThresholdBytes,
+                delegate,
+                SchedulerFactory.create(1, "ChangelogRetryScheduler", LOG));
+    }
+
+    BatchingStateChangeStore(
+            long persistDelayMs,
+            long sizeThresholdBytes,
+            StateChangeStore delegate,
+            ScheduledExecutorService scheduler) {
+        this.scheduleDelayMs = persistDelayMs;
+        this.scheduled = new LinkedList<>();
+        this.scheduler = scheduler;
+        this.sizeThresholdBytes = sizeThresholdBytes;
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void save(StoreTask storeTask) {
+        Collection<StateChangeSet> changeSets = storeTask.changeSets;
+        if (error.get() != null) {
+            LOG.debug("don't persist {} changesets, already failed", changeSets.size());
+            storeTask.fail(error.get());
+            return;
+        }
+        LOG.debug("persist {} changeSets", changeSets.size());
+        try {
+            synchronized (scheduled) {
+                scheduled.add(storeTask);
+                scheduledSizeInBytes += storeTask.getSize();
+                scheduleUploadIfNeeded();
+            }
+        } catch (Exception e) {
+            storeTask.fail(e);
+            throw e;
+        }
+    }
+
+    private void scheduleUploadIfNeeded() {
+        checkState(holdsLock(scheduled));
+        if (scheduledFuture.isDone() || isOverSizeThresholdAndCancellationSucceded()) {
+            scheduleUpload();
+        }
+    }
+
+    private boolean isOverSizeThresholdAndCancellationSucceded() {
+        return scheduledSizeInBytes >= sizeThresholdBytes && scheduledFuture.cancel(false);
+    }
+
+    private void scheduleUpload() {
+        checkState(scheduledFuture.isDone());
+        long delay = scheduleDelayMs;
+        if (scheduleDelayMs == 0 || scheduledSizeInBytes >= sizeThresholdBytes) {
+            scheduledFuture = scheduler.submit(this::drainAndSave);
+        } else {
+            scheduledFuture = scheduler.schedule(this::drainAndSave, delay, MILLISECONDS);
+        }
+    }
+
+    private void drainAndSave() {
+        Collection<StoreTask> tasks;
+        synchronized (scheduled) {
+            tasks = new ArrayList<>(scheduled);
+            scheduled.clear();
+            scheduledSizeInBytes = 0;
+        }
+        try {
+            if (error.get() != null) {
+                tasks.forEach(task -> task.fail(error.get()));
+                return;
+            }
+            delegate.save(tasks);
```

Comment:
This is a single-threaded upload now, I think we need more threads as discussed offline.

Reply:
Yes, each upload should be an independent file so it shouldn't affect the recovery.
I think we should have it in the first version (not sure whether this PR or not) especially if we ignore aborts.